nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-framework-cluster-protocol/src/main/java/org/apache/nifi/cluster/protocol/impl/SocketProtocolListener.java

```diff
@@ -134,15 +132,21 @@ public void dispatchRequest(final Socket socket) {
 
             // unmarshall message
             final ProtocolMessageUnmarshaller<ProtocolMessage> unmarshaller = protocolContext.createUnmarshaller();
-            final InputStream inStream = socket.getInputStream();
-            final CopyingInputStream copyingInputStream = new CopyingInputStream(inStream, maxMsgBuffer); // don't copy more than 1 MB
+            final ByteCountingInputStream countingIn = new ByteCountingInputStream(socket.getInputStream());
+            InputStream wrappedInStream = countingIn;
+            if (logger.isDebugEnabled()) {
+                final int maxMsgBuffer = 1024 * 1024;   // don't buffer more than 1 MB of the message
+                final CopyingInputStream copyingInputStream = new CopyingInputStream(wrappedInStream, maxMsgBuffer);
+                wrappedInStream = copyingInputStream;
+            }
 
             final ProtocolMessage request;
             try {
-                request = unmarshaller.unmarshal(copyingInputStream);
+                request = unmarshaller.unmarshal(wrappedInStream);
             } finally {
-                receivedMessage = copyingInputStream.getBytesRead();
                 if (logger.isDebugEnabled()) {
```

Comment:
Excellent!  I thought of using instanceof CopyingInputStream but didn't think it would help.  The potential race condition makes it useful and necessary.  I pushed a fix, will squash if needed.